R/verify.R

```diff
@@ -0,0 +1,53 @@
+#' Check Code Reproducibility
+#'
+#' Checks whether code in `teal_data` reproduces the stored objects.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' }
+#' @name verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#' @rdname verify
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
+  if (x@verified) {
+    return(x)
+  }
+  new_teal_data <- eval_code(teal_data(), x@code)
+
+  reproduced <- (inherits(new_teal_data, "teal_data") && isTRUE(all.equal(x@env, new_teal_data@env)))
+  if (reproduced) {
+    x@verified <- TRUE
+    methods::validObject(x)
+    x
+  } else {
+    stop("Code verification failed.")
```

Comment:
I love the fact that we return self in the verify allowing us to chain it.
The one place I can see improvement is the verification failure. Ideally, I would like to see the objects that failed verification so that I can correct them.

```r
td <- teal_data(iris = iris, code = "iris <- mtcars") |>
  within({
    mtcars <- mtcars
  }) |>
  verify()
#  Error in verify(within(teal_data(iris = iris, code = "iris <- mtcars"),  : 
#  Code verification failed.
```
Instead of that it would be nice to have something like:
```r
Unable to generate the object `iris` using code.
```

Reply:
You would have to split the condition on `reproduced` and handle `qenv.error` separately from `isTRUE(all.equal)`. Messages from `all.equal` can be passed to the error message.

Might be worth the trouble.

Reply:
Will think about it. I think it's cool suggestion

Reply:
I think we need to introduce hashes again. If all.equal fails, it just fails with a number of items in which environments differ.

```r
> new.env() -> a
> new.env() -> b
> 
> a$x <- 5
> b$x <- 10
> 
> all.equal(a, b)
[1] "Component “x”: Mean relative difference: 1"
```

We could also check whether names of objects are different, but in the above case they are the same `x`. I think the only way to found out if there are objects with the same name, but different structure is a lapply over objects and usage of all.equal or hashing objects.

Reply:
It quite clearly states which variable is not equal.

Reply:
Assuming the names of objects are the same across both `teal_data`s that we consider the verification could be like below

```
setMethod("verify", "teal_data", definition = function(x) {
  if (x@verified) {
    return(x)
  }
  new_teal_data <- eval_code(teal_data(), x@code)

  if (inherits(new_teal_data, "qenv.error")) {
    return(stop(new_teal_data@message, call. = FALSE))
  }

  reproduced <- isTRUE(all.equal(x@env, new_teal_data@env))
  if (reproduced) {
    x@verified <- TRUE
    methods::validObject(x)
    x
  } else {

    diff <- vapply(
      names(x@env),
      function(element) {
        isTRUE(all.equal(x@env[[element]], new_teal_data@env[[element]]))
      },
      logical(1)
    )

    stop("Code verification failed at object(s):", paste('\n', names(which(!diff)), collapse = '\n'))
  }
})
```

Need to figure out also the comparison/diff of object names in both `teal_data`s

Reply:
Interesting point. If vars in env are different to vars obtained by code, which one should take precedence? I mean if we are to point to an item that needs to be fixed, should we say "code failed to create x" or "x not provided"?

I am beginning to lean towards a generic message. Let the app dev sort it out themselves.

Reply:
I think we can inform about both. Check this PR d087e3b3ac968b52706aa1f8bb3cba22bf374e00
And this examples code

```r
x <- teal_data()
x <- within(x, {
  a <- 1
  b <- a+2
  c <- list(x = 2)
  d <- 5
})

new_teal_data <- teal_data()
new_teal_data <- within(new_teal_data, {
  a <- 1
  b <- a
  c <- list(x = 2)
})


reproduced <- isTRUE(all.equal(x@env, new_teal_data@env))
if (reproduced) {
  x@verified <- TRUE
  methods::validObject(x)
  x
} else {
  
  names_diff <- setdiff(names(x@env), names(new_teal_data@env))
  
  objects_diff <- vapply(
    intersect(names(x@env), names(new_teal_data@env)),
    function(element) {
      isTRUE(all.equal(x@env[[element]], new_teal_data@env[[element]]))
    },
    logical(1)
  )
  
  stop("Code verification failed at object(s):\n", paste(c(names_diff, names(which(!objects_diff))), collapse = '\n'))
}

```

# Result

```r
Error: Code verification failed at object(s):
d
b
```

Reply:
`return(stop())`? really?

Reply:
```
<qenv.error>@message
Error: trying to get slot "message" from an object (class "qenv.error") that is not an S4 object 
```

Reply:
But that's my point, why `setdiff(env, new)` and not vice versa?

Reply:
Ok, needed to change `qenv.error@message` to `conditionMessage(qenv.error)` ! 
Thanks, weird as this worked for dispatched get_code.qenv,error with `@message`

751b75ea2ece4bc1b22fbac5a0a8ca8fcd65a9e4

Reply:
It worked because you used the correct extract operator.

Reply:
of course, the code is longer, for `qenv.error` there was also 

```r
  if (inherits(new_teal_data, "qenv.error")) {
    stop(conditionMessage(new_teal_data), call. = FALSE)
  }

```

my first comment with the longer code aimed to compare only `teal_data` even though the verification for `qenv.error` was in the commit, but not in the comment

```r
new_teal_data <- eval_code(teal_data(), x@code)

  if (inherits(new_teal_data, "qenv.error")) {
    stop(conditionMessage(new_teal_data), call. = FALSE)
  }

  reproduced <- isTRUE(all.equal(x@env, new_teal_data@env))
  if (reproduced) {
    x@verified <- TRUE
    methods::validObject(x)
    x
  } else {
    names_diff <- setdiff(names(x@env), names(new_teal_data@env))

    objects_diff <- vapply(
      intersect(names(x@env), names(new_teal_data@env)),
      function(element) {
        isTRUE(all.equal(x@env[[element]], new_teal_data@env[[element]]))
      },
      logical(1)
    )

    stop("Code verification failed at object(s):\n", paste(c(names_diff, names(which(!objects_diff))), collapse = "\n"))
  }
  ```

Reply:
> of course, the code is longer, for `qenv.error` there was also
(...)
> my first comment with the longer code aimed to compare only `teal_data` even though the verification for `qenv.error` was in the commit, but not in the comment

I have no idea what you are talking about.

Reply:
removed my previous comment

Reply:
We keep talking about reproducible code, why the sudden change? :wink:

Reply:
I personally treat reproducible code actually as a reproducible environment. Probably a manner of philosophical point of view. Let's extend `setdiff(env, new)` with `setdiff(env, new)` but let's use different naming for the second case. For the latter we can say `code yeild object(s) not present in teal_data:` and here paste names of the other setdiff.

Reply:
Hey, added similar info for the other case, where code produces and object that does not exist in `teal_data` 47a3e0a5712e1fa8af8c6178c76c78f01d875018

<details><summary>R code</summary>
```r
x <- teal_data()
x <- within(x, {
  a <- 1
  b <- a+2
  c <- list(x = 2)
  d <- 5
})

new_teal_data <- teal_data()
new_teal_data <- within(new_teal_data, {
  a <- 1
  b <- a
  c <- list(x = 2)
  e <- 1
})

if (inherits(new_teal_data, "qenv.error")) {
  stop(conditionMessage(new_teal_data), call. = FALSE)
}

reproduced <- isTRUE(all.equal(x@env, new_teal_data@env))
if (reproduced) {
  x@verified <- TRUE
  methods::validObject(x)
  x
} else {
  names_diff <- setdiff(names(x@env), names(new_teal_data@env))
  
  objects_diff <- vapply(
    intersect(names(x@env), names(new_teal_data@env)),
    function(element) {
      isTRUE(all.equal(x@env[[element]], new_teal_data@env[[element]]))
    },
    logical(1)
  )
  
  names_diff_other <- setdiff(names(new_teal_data@env), names(x@env))
  
  error <- c(
    "\nCode verification failed at object(s):\n",
    paste(c(names_diff, names(which(!objects_diff))), collapse = "\n")
  )
  
  if (length(names_diff_other)) {
    error <- c(error,
              "\nObject(s) created with code that do not exist in teal_data:\n",
              paste(names_diff_other, collapse = "\n")
    )
  }
  
  stop(error)
}
```
</details>

```r
Error: 
Code verification failed at object(s):
d
b
Object(s) created with code that do not exist in teal_data:
e
```

Reply:
Ok, we ended up with

```r
a <- 1
b <- a+2
c <- list(x = 2)
d <- 5

tdata4 <- teal_data(
  a = a, b = b, c = c, d = d, 
  code = 
    "
    a <- 1
    b <- a
    c <- list(x = 2)
    e <- 1
    "
)

tdata4

# ✖ unverified teal_data object
# <environment: 0x000002d744cd9058> [L]
# Parent: <environment: package:teal.data>
# Bindings:
# • a: <dbl> [L]
# • b: <dbl> [L]
# • c: <named list> [L]
# • d: <dbl> [L]

verify(tdata4)

# Error in verify(tdata4) : Code verification failed.
# Object(s) recreated with code that have different structure in teal_data:
# • b
# Object(s) not created with code that exist in teal_data:
# • d
# Object(s) created with code that do not exist in teal_data:
# • e
```